Test Mode

chrisliebert opened the pull request:
There is some ambiguity in the current flight_mode design with direction, pin_num and value. The gpio library stores all of these in a pin object, but we Node has a Pin, a Direction and a Value.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/hgb2/psas-capstone/10)

<!-- Reviewable:end -->

Commit:
Added gpio test mode implementation skeleton.

Commit:
Added skeleton for I2C interface based on Tyler's changes to common/sensor.rs in /sensor_module_v1.

cwalexand commented:
Would you consider using the blank compatible gpio framework for test mode provided in the flt_mode_libs branch? However, this development isn't quite finished -- still needs to do some error handling which may take a couple days.

chrisliebert commented:
@cwalexand Yes, I did consider the flt_mode_libs branch. I believe this is a compatible, or I am not understanding your question. I still plan to change the init/update methods. If you compile test_mode for this branch there are many unused warnings.

chrisliebert commented:
Didn't mean to close the pull request, reopening.

cwalexand commented:
Okay, but flight mode is still changing. If test mode proceeds it, this will have to change later.

Commit:
merged /flt_mode_libs/src/test_mode/src/gpio.rs

chrisliebert commented:
I didn't realize that you had code in the test_mode directory on /flt_mode_libs. Merged gpio.rs.

Commit:
changed num from u64 to i32 in update function

cwalexand committed:
added gpio libraries

cwalexand committed:
implemented the control module design

cwalexand committed:
added control module comments

cwalexand committed:
updated control module design

cwalexand committed:
updated flight mode gpio library design

BrianBreniser commented on src/test_mode/src/gpio.rs:
```diff
@@ -1,8 +1,131 @@
+	    self.pins.push_back(node);
+    }
+
+	pub fn get_value(&mut self, pin_number: u64) -> u8 {
```
Any reason for this indentation?

BrianBreniser commented:
This has been here for a week now, are there any other comments? gpio.rs has formatting issues but short of that this looks good, and even there it's just some inconsistent indentation which is an easy fix. I think this needs tests but I'm willing to merge this right now and add tests later (Since I'm helping on this side now, I'm willing to write those tests)

Any other feedback here? If not I'll merge it Thursday.

cwalexand commented:
There is a compatible blank gpio template for test mode in the control_module_v1 branch that behaves identically to what is proposed for flight mode.  If you would prefer to use this, I will remove it from the control module branch.

cwalexand committed:
removed compatible gpio template for test mode

cwalexand committed:
resolved code review issues for gpio.rs:

1) changed linked list to vector
2) created an enum for Direction
3) removed the Node structure that contained pin number and direction

cwalexand committed:
resolved code review issues for control.rs:

1) changed state variable to from u8 to bool
2) used try! macro where appropriate
3) used absolute value function call for rate_x

cwalexand committed:
updated flight mode gpio design to use generic container, not linked list

Commit:
Merge remote-tracking branch 'remotes/origin/control_module_v1' into gpiotest_mode

Commit:
Adapted new GPIO interface and added a CMake build script that links libwrapper.so to JSBSim.

Commit:
Adapted new GPIO test interface.

Commit:
Updated JSBSim wrapper to match binder.rs.

Commit:
configured JSBSIM to work on a default linux installation with GCC-c++ 5

Commit:
Added cmake to cargo build.

Commit:
merged Sohail's prototype from slack

sanayani commented:
/me likey.  some thoughts:
if we want to hook up jsbsim at this point, we can replace the wrapper_init() call with one to fdm_create() & the wrapper_update() call with the one to fdm_get_version() & the wrapper_terminate() with fdm_close().  should be able to remove the wrapper_\* functions completely & just do them in rust.

will also need to update the .h file to include the extern "C" commands.  & take out the movie quote.  _lol_  s.

sanayani committed:
misc update

sanayani committed:
externed wrapper functions directly

sanayani committed:
update

Commit:
removed wrapper placeholder functions and removed delete statement

Commit:
Merge branch 'gpiotest_mode' of https://github.com/hgb2/PSAS-Capstone into gpiotest_mode

# Conflicts:
#	src/test_mode/src/JSBSimWrapper.cpp
#	src/test_mode/src/JSBSimWrapper.h

Commit:
removed unused enum.

chrisliebert commented:
There are more commits but they are no longer showing up here.

Commit:
change incorrect print statement

sanayani commented:
fdm isn't a u64.  its a pointer to an object instantiated by jsbsim's FGFDMExec class.  it needs to be set up as an opaque struct i.e.:

constructor:
pub fn fdm_create()->*mut libc::c_void;

function:
pub fn fdm_get_version(arg: *mut libc::c_void);

deconstructor:
pub fn fdm_close(arg: *mut libc::c_void);
s.

Commit:
removed the need for cmake to copy the static wrapper

Commit:
changed u64 to libc::c_void

Commit:
added call to binder::terminate: still a no-op instruction since pointer is not deleted

Commit:
Changed wrapper to static library, and called some actual JSBSim from rust.

Commit:
removed link directive from toml, it is not required.

sanayani committed:
removed pointer delete

pointer should be deleted at the rust level.  s.
